fix: reading single-task logs on Ray

alexjuda opened the pull request:
Marked as "draft" until the dependent PRs are merged:
* https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/49
* https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/50

# The problem

`orq task logs` (and the corresponding Python API) didn't work with reading task logs from Ray.

# This PR's solution

* Splits the awkward `RuntimeInterface.get_full_logs()` into to separate methods: `.get_workflow_logs()` and `.get_task_logs()` with non-optional arguments.
* Reimplements the log reading for Ray logs.
* Simplifies the log reading implementation for QE.

Main ticket: https://zapatacomputing.atlassian.net/browse/ORQSDK-678. Also solves:
- https://zapatacomputing.atlassian.net/browse/ORQSDK-750
- https://zapatacomputing.atlassian.net/browse/ORQSDK-584

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

alexjuda committed:
Filter ray logs by task run ID

Fix marking Ray log lines with task run IDs

Add a prototype impl of the task filtering

Split the log reader interface

Adopt the new log reader API

Update log tests

Tag logs with task inv ID

Fix missing wf inv ID in log struct

Add test for getting context IDs from Ray

Fix reading task logs for ray

Fix type errors

Add fluentbit todo

Document exceptions & add missing test for QERuntime

Add missing tests for ray runtime

Add missing tests to ray logs

Remove unneeded task ID map

Use lazy lines iterator

jamesclark-Zapata commented on src/orquestra/sdk/_ray/_ray_logs.py:
```diff
@@ -125,34 +121,33 @@ def _update_log_filenames(self):
-        # This is bad. The key should be a task invocation ID. To be fixed in the Jira
-        # ticket: https://zapatacomputing.atlassian.net/browse/ORQSDK-676
-        return {"logs": formatted}
+        return log_lines
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Can you explain why we return and yield? Whats the example usecase?

alexjuda replied:
You've found a leftover after the prev implementation, good catch! I'll update the PR, meanwhile see the motivation below.

The first approach from this PR read all lines from all files, returned them as a one big list, and the lines where being filtered only later on on another function.

The updated approach is a lazy sequence implemented as generator that yields line by line so don't have to keep big lists in memory only to dispose most of them. It's a memory optimization based on gut feeling. This is also similar to the implementation that was there before this PR, so I didn't want to make the performance terribly worse.

alexjuda committed:
Refactor yielding ray log lines & add more unit tests

Reimplement ray_logs

Use real dir layout in ray logs test data

Add test for iterating over log paths

Add test for iterating log lines

alexjuda committed:
Make log iter test robust to order changes

alexjuda committed:
Remove unused code

alexjuda committed:
Strip lines before asserting
